nifi-mock/src/main/java/org/apache/nifi/util/StandardProcessorTestRunner.java

```diff
@@ -221,28 +221,18 @@ public void run(final int iterations, final boolean stopOnFinish, final boolean
             } catch (final InterruptedException e1) {
             }
 
-            int finishedCount = 0;
-            boolean unscheduledRun = false;
             for (final Future<Throwable> future : futures) {
                 try {
                     final Throwable thrown = future.get(); // wait for the result
                     if (thrown != null) {
                         throw new AssertionError(thrown);
                     }
-
-                    if (++finishedCount == 1) {
-                        unscheduledRun = true;
-                        unSchedule();
-                    }
                 } catch (final Exception e) {
                 }
             }
 
-            if (!unscheduledRun) {
-                unSchedule();
-            }
-
             if (stopOnFinish) {
+                unSchedule();
```

Comment:
I'm not sure that I agree with this change. It is much simpler, but it changes the semantics a bit of how this runs. If the TestRunner is configured to use more than one thread and use multiple iterations, the old way would call `@OnUnscheduled` methods after the first thread completes, while other threads are potentially still running. This change makes `@OnUnscheduled` and `@OnStopped` methods functionally the same in the testing framework, while they are very different within NiFi. It's a fairly common mistake to see developers use `@OnUnscheduled` when they should use `@OnStopped`. So the code as-is would potentially highlight such issues in unit tests, while the change prevents it from catching this mistakes. I do see the need to check the `stopOnFinish` flag but perhaps it makes more sense to just update the `if` conditional to `if (++finishedCount == 1 && stopOnFinish) {`?

Reply:
@markap14 Thanks for your feedback and highlighting some aspects of how the test framework runs the processors and what the motivations are behind that. I totally agree that using `@OnUnscheduled` and `@OnStopped` in the proper way is a big issue that needs to be tested somehow. It was just weird for me (and broke my tests) that the second run of the processor happened in unscheduled state.
My test case:
- enqueue some FFs
- run (trigger) the processor once
- enqueue some more FFs
- run (trigger) the processor one more

I expected the processor to keep being scheduled between and during the 2 runs (as normally happens when I don't want to stop it).

I tried your suggestion about using `if (++finishedCount == 1 && stopOnFinish) {`. It works if I also add stopOnFinish check to `if (!unscheduledRun && stopOnFinish) {` (line 241 in the original version).
Actually, it falls back to my
```
            if (stopOnFinish) {
                unSchedule();
                stop();
            }
```
solution if run(1) was called.

Would that be fine with you?

Reply:
@markap14 I reveretd the change and added `stopOnFinish` checks to existing `if` statements.
Could you please check this version?